src/image/imageloadobserver.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * @license Copyright (c) 2003-2018, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module image/image/imageloadobserver
+ */
+
+import DomEventObserver from '@ckeditor/ckeditor5-engine/src/view/observer/domeventobserver';
+import DOMEmitterMixin from '@ckeditor/ckeditor5-utils/src/dom/emittermixin';
+
+/**
+ * Observes all new images added to the {@link module:engine/view/document~Document},
+ * fires {@link module:engine/view/document~Document#event:imageLoaded} and
+ * {@link module:engine/view/document~Document#layoutChanged} event every time when the new image
+ * has been loaded.
+ *
+ * **Note:** This event is not fired for images that has been added to the document and rendered as `complete` (already loaded).
+ *
+ * @extends module:engine/view/observer/domeventobserver~DomEventObserver
+ */
+export default class ImageLoadObserver extends DomEventObserver {
+	constructor( view ) {
+		super( view );
+
+		/**
+		 * List of img DOM elements that are observed by this observer.
+		 *
+		 * @private
+		 * @type {Set.<HTMLElement>}
+		 */
+		this._observedElements = new Set();
+
+		/**
+		 * DOM emitter used for listening images `load` event.
+		 *
+		 * @private
+		 * @type {module:utils/dom/emittermixin~DomEmitterMixin}
+		 */
+		this._domEmitter = Object.create( DOMEmitterMixin );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	observe( domRoot, name ) {
+		const viewRoot = this.document.getRoot( name );
+
+		viewRoot.on( 'change:children', ( evt, node ) => {
+			this.view.once( 'render', () => {
+				if ( !node.is( 'element' ) || node.is( 'attributeElement' ) ) {
+					return;
+				}
+
+				const domNode = this.view.domConverter.mapViewToDom( node );
+
+				// If there is no `domNode` it means that it was removed from the DOM in the meanwhile.
+				if ( !domNode ) {
+					return;
+				}
+
+				for ( const domElement of domNode.querySelectorAll( 'img' ) ) {
+					if ( !this._observedElements.has( domElement ) ) {
+						this._domEmitter.listenTo( domElement, 'load', ( evt, domEvt ) => this.onDomEvent( domEvt ) );
+						this._observedElements.add( domElement );
+					}
+				}
+
+				// Clean up the list of observed elements from elements that has been removed from the root.
+				for ( const domElement of this._observedElements ) {
+					if ( !domRoot.contains( domElement ) ) {
+						this._domEmitter.stopListening( domElement );
+						this._observedElements.delete( domElement );
+					}
+				}
+			} );
+		} );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	onDomEvent( domEvent ) {
+		this.document.fire( 'layoutChanged' );
+		this.fire( 'imageLoaded', domEvent );
```

Comment:
On `this`? Why not on the document?

Reply:
https://github.com/ckeditor/ckeditor5-engine/blob/dfa0b39caea05b6df984c6a49949091c8186557c/src/view/observer/domeventobserver.js#L94-L97

Reply:
Uh, I forgot that this is confusing.

Reply:
https://github.com/ckeditor/ckeditor5-engine/issues/1448 RFC